### ISSUE TITLE
rustdoc overflows stack on latest nightly

### ISSUE THREAD

**Issue by benesch** [rust-lang/rust#59502]
To reproduce:

```bash
$ git clone git@github.com:rust-num/num-traits.git
# Check out a known bad commit, just in case.
$ git checkout 8915b74ae4f4fe42fae329852a2e24934c0623d5
$ cargo +nightly doc
thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
error: Could not document `num-traits`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name num_traits src/lib.rs --color always -o /Users/benesch/Sites/materialize/num-traits/target/doc --cfg 'feature="default"' --cfg 'feature="std"' -L dependency=/Users/benesch/Sites/materialize/num-traits/target/debug/deps --cfg has_i128` (signal: 6, SIGABRT: process abort signal)
```

Rustdoc version:

```
rustdoc 1.35.0-nightly (33ef0bad2 2019-03-27)
```

I can reproduce on both macOS and Linux. Let me know if there's anything else I can provide!

**Comment by akinnane** [comment]
I'm getting the same error on Linux using the same rustdoc version.

**Comment by jonas-schievink** [comment]
Can you try obtaining a backtrace with gdb or lldb?

**Comment by eddyb** [comment]
Someone on IRC got this in a backtrace:
```
get_real_types () at src/librustdoc/clean/mod.rs:1820
```

I'm trying to do a review of #59004, I suspect it needs a bit more work to be robust.

**Comment by GuillaumeGomez** [comment]
@eddyb I let you take care of it?

**Comment by eddyb** [comment]
@GuillaumeGomez @QuietMisdreavus Can you revert #59004 entirely until I have the time to look into another minor refactor of rustdoc? It might be a few days (and that if I don't forget about it).

Or at least comment out the calls to the recursive functions that cause the stack overflow.

**Comment by GuillaumeGomez** [comment]
Sure.

**Comment by GuillaumeGomez** [comment]
I sent a PR and answered a few of your questions. The main point of what I added was to generate a list of names, not of types. Therefore, I don't care if I have two types with the same name. It doesn't matter. :)

**Comment by Alexendoo** [comment]
Fixed by #59539

**Comment by camelid** [comment]
Note: I have since added a regression test for this issue, in 908a9d4aad61f4b8108ff91ae42f76e67dec5148.
